File: src/utils/taskRouting.ts
export const TASK_TYPE_LABELS: Record<string, string> = {
  "chapter.draft": "生成草稿",
  "chapter.continue": "续写",
  "chapter.rewrite": "改写",
  "chapter.plan": "章节计划",
  "prose.naturalize": "去 AI 味",
  "character.create": "创建角色卡",
  "world.create_rule": "创建世界规则",
  "plot.create_node": "创建剧情节点",
  "glossary.create_term": "创建名词",
  "narrative.create_obligation": "创建叙事义务",
  "timeline.review": "时间线审阅",
  "relationship.review": "关系审阅",
  "dashboard.review": "仪表盘审阅",
  "export.review": "导出审阅",
  "consistency.scan": "一致性检查",
  "blueprint.generate_step": "生成蓝图步骤",
  custom: "自定义任务",
};

export const TASK_ROUTE_OPTIONS = [
  "chapter.draft",
  "chapter.continue",
  "chapter.rewrite",
  "chapter.plan",
  "prose.naturalize",
  "character.create",
  "world.create_rule",
  "consistency.scan",
  "blueprint.generate_step",
  "plot.create_node",
  "glossary.create_term",
  "narrative.create_obligation",
  "timeline.review",
  "relationship.review",
  "dashboard.review",
  "export.review",
  "custom"
].map(value => ({ value, label: TASK_TYPE_LABELS[value] || value }));


export interface EditorAiAction {
  taskType: string;
  label: string;
  category: "writing" | "character" | "world" | "plot" | "review";
}
export type EditorAiCategory = EditorAiAction["category"];

// WP-5: 编辑器固定 9 按钮任务清单（canonical）
export const EDITOR_AI_ACTIONS: EditorAiAction[] = [
  { taskType: "chapter.continue", label: "续写章节", category: "writing" },
  { taskType: "chapter.draft", label: "生成章节草稿", category: "writing" },
  { taskType: "chapter.plan", label: "生成章节计划", category: "writing" },
  { taskType: "chapter.rewrite", label: "改写选区", category: "writing" },
  { taskType: "prose.naturalize", label: "去 AI 味", category: "writing" },
  { taskType: "character.create", label: "创建角色卡", category: "character" },
  { taskType: "world.create_rule", label: "创建世界规则", category: "world" },
  { taskType: "plot.create_node", label: "创建剧情节点", category: "plot" },
  { taskType: "consistency.scan", label: "一致性扫描", category: "review" },
];

export const EDITOR_AI_TASK_TYPES = EDITOR_AI_ACTIONS.map((action) => action.taskType);
export const EDITOR_AI_TASK_SET = new Set(EDITOR_AI_TASK_TYPES);


export const DIFF_TASK_TYPES = new Set(["chapter.rewrite", "prose.naturalize"]);

const TASK_REQUIREMENT_MAP: Record<string, {
  requiresChapterId: boolean;
  requiresSelectedText: boolean;
  requiresUserInstruction: boolean;
  requiresChapterContent: boolean;
}> = {
  "chapter.draft": {
    requiresChapterId: true,
    requiresSelectedText: false,
    requiresUserInstruction: false,
    requiresChapterContent: false
  },
  "chapter.continue": {
    requiresChapterId: true,
    requiresSelectedText: false,
    requiresUserInstruction: false,
    requiresChapterContent: false
  },
  "chapter.plan": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: false,
    requiresChapterContent: false
  },
  "chapter.rewrite": {
    requiresChapterId: true,
    requiresSelectedText: true,
    requiresUserInstruction: false,
    requiresChapterContent: false
  },
  "prose.naturalize": {
    requiresChapterId: true,
    requiresSelectedText: true,
    requiresUserInstruction: false,
    requiresChapterContent: false
  },
  "character.create": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: true,
    requiresChapterContent: false
  },
  "world.create_rule": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: true,
    requiresChapterContent: false
  },
  "plot.create_node": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: true,
    requiresChapterContent: false
  },
  "consistency.scan": {
    requiresChapterId: true,
    requiresSelectedText: false,
    requiresUserInstruction: false,
    requiresChapterContent: true
  },
  "glossary.create_term": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: true,
    requiresChapterContent: false
  },
  "narrative.create_obligation": {
    requiresChapterId: false,
    requiresSelectedText: false,
    requiresUserInstruction: true,
    requiresChapterContent: false
  },
};

export function getTaskRequirements(taskType: string) {
  const canonical = taskType.trim();
  return (
    TASK_REQUIREMENT_MAP[canonical] || {
      requiresChapterId: false,
      requiresSelectedText: false,
      requiresUserInstruction: false,
      requiresChapterContent: false
    }
  );
}

export function isEditorAiTask(taskType: string): boolean {
  return EDITOR_AI_TASK_SET.has(taskType.trim());
}

File: src-tauri/src/services/task_routing.rs
use serde::Serialize;
use std::borrow::Cow;

pub const CORE_TASK_ROUTE_TYPES: &[&str] = &[
    "chapter.draft",
    "chapter.continue",
    "chapter.rewrite",
    "chapter.plan",
    "prose.naturalize",
    "character.create",
    "world.create_rule",
    "consistency.scan",
    "blueprint.generate_step",
    "plot.create_node",
    "glossary.create_term",
    "narrative.create_obligation",
    "timeline.review",
    "relationship.review",
    "dashboard.review",
    "export.review",
];

pub const TASK_ROUTE_TYPES_WITH_CUSTOM: &[&str] = &[
    "chapter.draft",
    "chapter.continue",
    "chapter.rewrite",
    "chapter.plan",
    "prose.naturalize",
    "character.create",
    "world.create_rule",
    "consistency.scan",
    "blueprint.generate_step",
    "plot.create_node",
    "glossary.create_term",
    "narrative.create_obligation",
    "timeline.review",
    "relationship.review",
    "dashboard.review",
    "export.review",
    "custom",
];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StoryAuthorityLayer {
    StoryConstitution,
    FormalAssets,
    SceneExecution,
    ReviewAudit,
    Custom,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StoryStateLayer {
    Constitution,
    Asset,
    DynamicScene,
    Review,
    Custom,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ReviewGateMode {
    ManualRequired,
    ManualRecommended,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskExecutionContract {
    pub authority_layer: StoryAuthorityLayer,
    pub state_layer: StoryStateLayer,
    pub capability_pack: &'static str,
    pub review_gate: ReviewGateMode,
}

pub fn canonical_task_type<'a>(task_type: &'a str) -> Cow<'a, str> {
    let task_type = task_type.trim();
    match task_type {
        "chapter_draft" | "generate_chapter_draft" | "draft" | "chapter.draft" => {
            Cow::Borrowed("chapter.draft")
        }
        "chapter_continue" | "continue_chapter" | "continue_draft" | "chapter.continue" => {
            Cow::Borrowed("chapter.continue")
        }
        "chapter_rewrite" | "rewrite_selection" | "chapter.rewrite" => {
            Cow::Borrowed("chapter.rewrite")
        }
        "chapter_plan" | "plan_chapter" | "chapter.plan" => Cow::Borrowed("chapter.plan"),
        "prose_naturalize" | "deai_text" | "prose.naturalize" => Cow::Borrowed("prose.naturalize"),
        "character_create" | "character.create" => Cow::Borrowed("character.create"),
        "world.generate" | "world_create_rule" | "world.create_rule" => {
            Cow::Borrowed("world.create_rule")
        }
        "plot.generate" | "plot_create_node" | "plot.create_node" => {
            Cow::Borrowed("plot.create_node")
        }
        "glossary_create_term" | "glossary.create" | "glossary.create_term" => {
            Cow::Borrowed("glossary.create_term")
        }
        "narrative_create_obligation" | "narrative.create" | "narrative.create_obligation" => {
            Cow::Borrowed("narrative.create_obligation")
        }
        "timeline_review" | "timeline.scan" | "timeline.review" => Cow::Borrowed("timeline.review"),
        "relationship_review" | "relationships.review" | "relationship.review" => {
            Cow::Borrowed("relationship.review")
        }
        "dashboard_review" | "dashboard.analyze" | "dashboard.review" => {
            Cow::Borrowed("dashboard.review")
        }
        "export_review" | "export.check" | "export.review" => Cow::Borrowed("export.review"),
        "scan_consistency" | "consistency_scan" | "consistency.scan" => {
            Cow::Borrowed("consistency.scan")
        }
        "generate_blueprint_step" | "blueprint_generate" | "blueprint.generate_step" => {
            Cow::Borrowed("blueprint.generate_step")
        }
        "custom" => Cow::Borrowed("custom"),
        _ => Cow::Borrowed(task_type),
    }
}

pub fn is_core_task_type(task_type: &str) -> bool {
    let canonical = canonical_task_type(task_type);
    CORE_TASK_ROUTE_TYPES.contains(&canonical.as_ref())
}

pub fn task_execution_contract(task_type: &str) -> TaskExecutionContract {
    let canonical = canonical_task_type(task_type);
    match canonical.as_ref() {
        "blueprint.generate_step" => TaskExecutionContract {
            authority_layer: StoryAuthorityLayer::StoryConstitution,
            state_layer: StoryStateLayer::Constitution,
            capability_pack: "blueprint-planning-pack",
            review_gate: ReviewGateMode::ManualRecommended,
        },
        "character.create"
        | "world.create_rule"
        | "plot.create_node"
        | "glossary.create_term"
        | "narrative.create_obligation" => TaskExecutionContract {
            authority_layer: StoryAuthorityLayer::FormalAssets,
            state_layer: StoryStateLayer::Asset,
            capability_pack: "asset-building-pack",
            review_gate: ReviewGateMode::ManualRecommended,
        },
        "chapter.draft" | "chapter.continue" | "chapter.rewrite" | "chapter.plan"
        | "prose.naturalize" => TaskExecutionContract {
            authority_layer: StoryAuthorityLayer::SceneExecution,
            state_layer: StoryStateLayer::DynamicScene,
            capability_pack: "scene-production-pack",
            review_gate: ReviewGateMode::ManualRequired,
        },
        "consistency.scan"
        | "timeline.review"
        | "relationship.review"
        | "dashboard.review"
        | "export.review" => TaskExecutionContract {
            authority_layer: StoryAuthorityLayer::ReviewAudit,
            state_layer: StoryStateLayer::Review,
            capability_pack: "review-guard-pack",
            review_gate: ReviewGateMode::ManualRequired,
        },
        _ => TaskExecutionContract {
            authority_layer: StoryAuthorityLayer::Custom,
            state_layer: StoryStateLayer::Custom,
            capability_pack: "custom-pack",
            review_gate: ReviewGateMode::ManualRequired,
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonicalize_legacy_aliases() {
        assert_eq!(
            canonical_task_type("generate_chapter_draft"),
            "chapter.draft"
        );
        assert_eq!(canonical_task_type("scan_consistency"), "consistency.scan");
        assert_eq!(
            canonical_task_type("generate_blueprint_step"),
            "blueprint.generate_step"
        );
        assert_eq!(
            canonical_task_type("glossary.create"),
            "glossary.create_term"
        );
        assert_eq!(
            canonical_task_type("narrative_create_obligation"),
            "narrative.create_obligation"
        );
        assert_eq!(
            canonical_task_type("relationships.review"),
            "relationship.review"
        );
        assert_eq!(canonical_task_type("timeline.scan"), "timeline.review");
        assert_eq!(canonical_task_type("dashboard.analyze"), "dashboard.review");
        assert_eq!(canonical_task_type("export.check"), "export.review");
    }

    #[test]
    fn canonicalize_keeps_unknown() {
        assert_eq!(canonical_task_type("my.custom.skill"), "my.custom.skill");
    }

    #[test]
    fn core_task_type_check() {
        assert!(is_core_task_type("chapter_draft"));
        assert!(is_core_task_type("chapter.draft"));
        assert!(!is_core_task_type("custom"));
        assert!(!is_core_task_type("my.custom.skill"));
    }

    #[test]
    fn execution_contract_for_scene_task() {
        let contract = task_execution_contract("chapter.rewrite");
        assert_eq!(
            contract.authority_layer,
            StoryAuthorityLayer::SceneExecution
        );
        assert_eq!(contract.state_layer, StoryStateLayer::DynamicScene);
        assert_eq!(contract.capability_pack, "scene-production-pack");
        assert_eq!(contract.review_gate, ReviewGateMode::ManualRequired);
    }

    #[test]
    fn execution_contract_for_asset_task() {
        let contract = task_execution_contract("character.create");
        assert_eq!(contract.authority_layer, StoryAuthorityLayer::FormalAssets);
        assert_eq!(contract.state_layer, StoryStateLayer::Asset);
        assert_eq!(contract.capability_pack, "asset-building-pack");
        assert_eq!(contract.review_gate, ReviewGateMode::ManualRecommended);
    }

    #[test]
    fn execution_contract_for_unknown_task() {
        let contract = task_execution_contract("my.custom.skill");
        assert_eq!(contract.authority_layer, StoryAuthorityLayer::Custom);
        assert_eq!(contract.state_layer, StoryStateLayer::Custom);
        assert_eq!(contract.capability_pack, "custom-pack");
        assert_eq!(contract.review_gate, ReviewGateMode::ManualRequired);
    }
}
